packages/salesforcedx-apex-replay-debugger/src/adapter/apexReplayDebug.ts

```diff
@@ -289,10 +324,19 @@ export class ApexReplayDebug extends LoggingDebugSession {
     this.sendResponse(response);
   }
 
-  protected async scopesRequest(
+  public async scopesRequest(
     response: DebugProtocol.ScopesResponse,
     args: DebugProtocol.ScopesArguments
   ): Promise<void> {
+    try {
+      if (this.logContext.hasHeapDumpForTopFrame()) {
```

Comment:
When VS Code asks for variables, it always asks for the scopes first. By this time, we have a call stack and variables from the debug log. So, we check if there’s a heapdump for the topmost frame and use that heapdump for variables. The variables from the debug log get copied by value before they are replaced by variables from heapdump.

It doesn’t verify that the current stopped line has a checkpoint. The user may have removed the checkpoint from the client, but not upload the changes to the server, so I don’t know for sure. Open for discussion.

Reply:
At the point you start debugging, checkpoints (in VS Code or on the server) are 100% moot and the only thing that matters are the HEAP_DUMP entries in the logs. The fact is, the user didn't have to use VS Code to set checkpoints to create the logs, that can still all be done server side. They only things they really need to use the replay-debugger is to have the code locally and a log file.